### ISSUE TITLE
Type invalid due to import, only on alternating runs

### ISSUE THREAD

**Issue by jvanasco** [python/mypy#17202]
**Bug Report**

I think this is a bug, as I get inconsistent results across multiple invocations.

Depending on how I define a Type Alias, it can or can not be used as a valid type in another file - on every other invocation of mypy.

```python
from typing import Literal
from typing import Optional
from typing import Union

TYPE_FOO = Optional[str]
TYPE_BAR = Union[str, None]
    
def biz(a: TYPE_FOO, b: TYPE_BAR) -> Literal[True]:
    return True
```

However, if the type aliases are split into another file, `TYPE_FOO` (using `Optional`) is sometimes invalid in mypy and triggers "Variable "TYPE_FOO" is not valid as a type [valid-type]" – while the functionally equivalent `TYPE_BAR` (using `Union`) is always valid.  The error from `Optional` seems to appear on every-other invocation of mypy.  I can't seem to recreate the error reliably with the small test case alone, but dropping it into larger projects seems to trigger it on every-other run.

  
**To Reproduce**

`custom_types.py`

```python
from typing import Optional
from typing import Union

TYPE_FOO = Optional[str]
TYPE_BAR = Union[str, None]
```

`project.py`

```python
from typing import Literal
from .custom_types import TYPE_FOO
from .custom_types import TYPE_BAR

def biz(a: TYPE_FOO, b: TYPE_BAR) -> Literal[True]:
    return True
```

**Expected Behavior**

I expected no errors for either usage.
I expected consistent behavior from mypy across runs when it thinks there is an issue.

**Actual Behavior**

MyPy alternates an error for the imported type that uses `Optional` on every-other run:

    Variable "TYPE_FOO" is not valid as a type [valid-type]
    
i.e. The first run has no errors, the second has the error, the third has no errors, the fourth has the error, etc

The type using `Union` never triggers an error.

**Your Environment**

- Mypy version used:
    1.10
- Mypy command-line flags:
    NA
- Mypy configuration options from `mypy.ini` (and other config files):
    check_untyped_defs = True
- Python version used:
    3.10

**Comment by lukecivantos** [comment]
Something similar is happening to me. Every other time I run mypy it runs on my `.venv` even with `exclude = ^\.venv/`
